text/0000-unsafe-attributes.md

```diff
@@ -0,0 +1,175 @@
+- Feature Name: `unsafe_attributes`
+- Start Date: 2022-10-11
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Consider some attributes 'unsafe', so that they must only be used like this:
+
+```rust
+#[unsafe(no_mangle)]
+```
+
+# Motivation
+[motivation]: #motivation
+
+Some of our attributes, such as `no_mangle`, can be used to
+[cause Undefined Behavior without any `unsafe` block](https://github.com/rust-lang/rust/issues/28179).
+If this was regular code we would require them to be placed in an `unsafe {}`
+block, but since they are attributes that makes less sense. Hence we need a
+concept of 'unsafe attributes' and accompanying syntax to declare that one is
+aware of the UB risks here (and it might be good to add a SAFETY comment
+explaining why this use of the attribute is fine).
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+*Example explanation for `no_mangle`; the other attributes need something similar.*
+
+When declaring a function like this
+
+```rust
+#[no_mangle]
+pub fn write(...) { ... }
+```
+
+this will cause Rust to generate a globally visible function with the
+linker/export name `write`. As consequence of that, other code that wants to
+call the
+[C `write` function](https://man7.org/linux/man-pages/man2/write.2.html) might
+end up calling this other `write` instead. This can easily lead to Undefined
+Behavior:
+- The other `write` might have the wrong signature, so arguments are passed
+  incorrectly.
+- The other `write` might not have the expected behavior of
+  [write](https://man7.org/linux/man-pages/man2/write.2.html), causing code
+  relying on this behavior to misbehave.
+
+To avoid this, when declaring a function `no_mangle`, it is important that the
+name of the function does not clash with other globally named functions. Similar
+to how `unsafe { ... }` blocks are used to acknowledge that this code is
+dangerous and needs manual checking, `unsafe(no_mangle)` acknowledges that
+`no_mangle` is dangerous and needs to be manually checked for correctness:
+
+```rust
+// SAFETY: there is no other global function of this name
+#[unsafe(no_mangle)]
+pub fn my_own_write(...) { ... }
+```
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Some attributes (e.g. `no_mangle`, `export_name`, `link_section` -- see
+[here](https://github.com/rust-lang/rust/issues/82499) for a more complete list)
+are considered "unsafe" attributes. An unsafe attribute must only be used inside
+`unsafe(...)` in the attribute declaration, like
+
+```rust
+#[unsafe(no_mangle)]
+```
+
+For backwards compatibility reasons, using these attributes outside of
+`unsafe(...)` is just a lint, not a hard error. Initially, this lint will be
```

Comment:
Maybe give it a temporary name for clarity and to distinguish it from the `unsafe_code` one which is talked about below? e.g. `unsafe_attr_outside_unsafe`.

This also allows to give an example for users on how to prepare their code for the future, e.g.. `#![deny(unsafe_attr_outside_unsafe)]`.